### PR TITLE
Battle 2k3: Add negative attribute modifiers and allow state infliction by skills outside of battles

### PR DESCRIPTION
**Pull request by rueter37**
This PR implements the negative attribute modifiers used in RPG Maker 2003 which allows for damage absorbing (e.g. dragons absorbing fire-type skills) and the possibility for skills to inflict states outside of battles (this makes the "Regenerationstrank" in "Vampires Dawn II" work outside of battles). By the way, negative attribute modifiers even allow for "funny" things like taking damage through healing skills (but you cannot get killed this way, 1 HP remain - tested with RPG_RT). This is probably my "most insane" PR so I expect many changes needed. Feedback is very appreciated!

### CODE REVIEW

**Comment by fmatthew5876** [comment]
I had a thought about this, let me know what you think.

I find this whole idea of `GetAffectedHp() < 0` meaning hp is not affected confusing. At a minimum we should wrap this in a function call and stop doing `< 0` inline everywhere.

My idea:

What if we had boolean flags for hp, sp, atk, def, spi, and agi to say whether or not the skill affects them, and then just use + or -  signed numbers to indicate healing vs damage for each one. Then there is no more `healing` or `IsPositive()` flag. Such a design also allows custom battle systems to extend and add skills which can affect each parameter individually.

I've always had an idea to try this approach, but haven't yet, so I don't know for sure if it'll end up better or worse.

File: src/game_battlealgorithm.cpp
Comment (fmatthew5876) [inline src/game_battlealgorithm.cpp]: Since you're refactoring and testing here, this min/max junk should be replaced with `Utils::Clamp()`

Comment (rueter37) [reply src/game_battlealgorithm.cpp]: Did the replacements now.

**Comment by rueter37** [comment]
Your idea sounds interesting. Being able to affect each parameter individually would be a nice addition for #2219. But I would like to hear what the others think about it.

**Comment by fmatthew5876** [comment]
Regardless, I think the overall approach you took here probably makes sense, Any further enhancement can be a later PR.

Please also test carefully the 2k battle system. It's message based state machine heavily relies on the various battle algorithm flags

**Comment by rueter37** [comment]
Did tests with the games "Vampires Dawn" (RPG Maker 2000), "Vampires Dawn II" (RPG Maker 2003) and some custom test cases and did not encounter any regressions.

**Review by fdelapena — APPROVED** [review]

**Comment by fmatthew5876** [comment]
I saw this in the RE code, couldn't believe it, but then confirmed it's true...

There is a *massive* RPG_RT bug with negative attributes. ❗ 

Negative attribute HP healing is doubled in it's effect!

To see this, Create a skill that does exactly 50 damage, with fire element. Create an actor with -100% fire attribute.

Now start a battle with actor at 1 hp, and  have an enemy use the fire skill on the actor. 50 hp will be displayed in green pop up numbers, but actor HP will actually increase by 100!

The reason for this is because in the battle code, first they call `RPG_Actor::ApplyHpDamage()` unconditionally with a negative value. The function is meant to handle hp damage, but because the amount is negative, it actually recovers HP. Then it starts to spawn white popup numbers with a negative value.

Then they immediately check if the damage is negative and call `RPG_Actor::ApplyHpRecovery()` with the value flipped. This heals the actor again, and spawns green numbers with the effect amount, immediately overwriting the previous white numbers.

Here is the function with this bug ...

| Address | Class | Name |
| -- | -- | -- |
| 0049CFE8 | RPG_SceneBattle | ApplyHpDamageEffect |

**Comment by rueter37** [comment]
The "double strength for negative effects" bug is emulated now.

**Comment by rueter37** [comment]
@fmatthew5876 To avoid conflicts with your PR (#2320) I made a separate branch which implements the negative attribute modifiers on top of #2320. This task is already completed and my tests worked so far. So we can wait until #2320 is merged then I will force-push my new changes into this PR.

**Comment by fmatthew5876** [comment]
You can rebase your PR now onto master, #2320 was merged. Once that's done I'll test and review this guy.

**Comment by rueter37** [comment]
Did the rebase now, you can test and review it now.

File: src/game_battlealgorithm.cpp
Comment (fmatthew5876) [inline src/game_battlealgorithm.cpp:935]: Even if it's not used in `Normal`, we should for consistency set the `negative_effect` flag here in case something later wants to rely on it.

Also this `*2` to emulate the bug is not correct. The double affect happens only for HP, but not other stats. So if you want to emulate the bug you'll need to double the hp effect but not others.

Comment (rueter37) [reply src/game_battlealgorithm.cpp:935]: I don't think we should emulate the animated damage number bug for better user experience in my opinion.

File: src/game_battlealgorithm.cpp
Comment (fmatthew5876) [inline src/game_battlealgorithm.cpp:1352]: This is one part where the code gets super confusing, when I first read this I thought you were reversing the attribute shift but you're not because both of these flags get set at the same time.

This inline xor everywhere is super ugly. Can you please wrap this in an inline helper function with a useful name so we can know whats going on.

Btw for this stuff you can push new commits instead of rebasing if thats easier.

File: src/game_battlealgorithm.cpp
Comment (fmatthew5876) [reply src/game_battlealgorithm.cpp:935]: I think I agree. We'll probably get "bug reports" about the effect being double, but we'll then just have to point out it's actually RPG_RT compatible!

Comment (rueter37) [reply src/game_battlealgorithm.cpp:935]: The double affect bug is now applied on HP only.

File: src/game_battlealgorithm.cpp
Comment (rueter37) [reply src/game_battlealgorithm.cpp]: Moved the XOR-Operations into own functions.

File: src/game_battlealgorithm.cpp
Comment (fmatthew5876) [inline src/game_battlealgorithm.cpp:943]: One more edge case, the double effect bug *does not* happen for absorb hp. Only normal hp damage..

File: src/game_battlealgorithm.cpp
Comment (fmatthew5876) [inline src/game_battlealgorithm.cpp:1169]: One more edge case, the double effect bug *does not* happen for absorb hp. Only normal hp damage..

File: src/game_battlealgorithm.cpp
Comment (fmatthew5876) [inline src/game_battlealgorithm.cpp:1188]: Confirmed this looks right, HP recovery code clamps (1, maxhp), so a negative healing effect cannot kill.

Pretty stupid really. You can't make a game that you can kill undead with healing spells, only weaken them to 1 hp..

File: src/game_battlealgorithm.cpp
Comment (fmatthew5876) [inline src/game_battlealgorithm.cpp:1201]: Don't you still need to reduce the SP recovery by the `sp_cost` here?

File: src/game_battler.cpp
Comment (fmatthew5876) [inline src/game_battler.cpp:283]: Not sure how we missed this last time. You're right that this case allows you to inflict battle only states, even from the menu.

**Comment by fmatthew5876** [comment]
This looks good to me. In particular I really like how `Skill::Execute` has changed from 2 overly long branches for positive and negative, to now flatter code which has a separate branch for each effect. It's more readable.

Once you address the few minor nits remaining, I can approve this.

File: src/game_battlealgorithm.cpp
Comment (fmatthew5876) [inline src/game_battlealgorithm.cpp:1185]: All these Clamps are written wrong, it should be (effect, 0, maxvalue)

File: src/game_battlealgorithm.cpp
Comment (rueter37) [reply src/game_battlealgorithm.cpp:1170]: Fixed.

File: src/game_battlealgorithm.cpp
Comment (rueter37) [reply src/game_battlealgorithm.cpp:1202]: Well, the code as is works for me. The else branch is taking SP damage from the healing skill because of the negative effect and in my tests it worked correctly.

File: src/game_battlealgorithm.cpp
Comment (rueter37) [reply src/game_battlealgorithm.cpp:1186]: This is probably a dumb question but what value I'm supposed to use for maxvalue? Is it MaxDamageValue() or the maximum positive integer value?

**Comment by rueter37** [comment]
I have added a commit which addresses two special cases which I must have overseen:
- If the source tries to absorb HP from a target and this fails due to negative damage, the amount healed of the target is no longer restricted by the targets current HP.
- If the target recovers SP due to negative damage, the amount recovered is no longer restricted by the targets current SP.

File: src/game_battlealgorithm.cpp
Comment (fmatthew5876) [reply src/game_battlealgorithm.cpp:1206]: No you're right. This was a mistake on my part

File: src/game_battlealgorithm.cpp
Comment (fmatthew5876) [reply src/game_battlealgorithm.cpp:1186]: The maxvalue depends on the context. You just need to flip the order of the arguments here. While what you wrote technically works, it doesn't read well.

The function signature is `Clamp(value, min, max)`

So here it would be:

```
Clamp(effect, 0, GetTarget()->GetMaxHp() - GetTarget()->GetHp())
```

Comment (rueter37) [reply src/game_battlealgorithm.cpp]: The order of the arguments have been corrected now.

**Review by fmatthew5876 — APPROVED** [review]
Looks good to me now, thanks!

**Comment by fmatthew5876** [comment]
This one needs a rebase again

**Comment by Ghabry** [comment]
sorry, this will be the next one that gets in :)

**Comment by rueter37** [comment]
Did the rebase now.

**Comment by Ghabry** [comment]
In general I let @fmatthew5876 handle any PRs that are algorithm related. I don't really want to dive into this

**Pull request merged**
